modules/core/src/utils/logtagconfigparser.cpp

```diff
@@ -0,0 +1,280 @@
+        return;
+    }
+    const size_t firstNonWildcard = name.find_first_not_of("*.");
+    if (hasPrefixWildcard && firstNonWildcard == npos)
+    {
+        m_parsedGlobal.level = level;
+        return;
+    }
+    const bool hasSuffixWildcard = (name[len - 1u] == '*');
+    const size_t lastNonWildcard = name.find_last_not_of("*.");
+    std::string trimmedNamePart = name.substr(firstNonWildcard, lastNonWildcard - firstNonWildcard + 1u);
+    // The case of a single asterisk has been handled above;
+    // here we only handle the explicit use of "global" in the log config string.
+    const bool isGlobal = (trimmedNamePart == "global");
+    if (isGlobal)
+    {
+        m_parsedGlobal.level = level;
+        return;
+    }
+    LogTagConfig result(trimmedNamePart, level, false, hasPrefixWildcard, hasSuffixWildcard);
+    if (hasPrefixWildcard)
+    {
+        m_parsedAnyPart.emplace_back(std::move(result));
+    }
+    else if (hasSuffixWildcard)
+    {
+        m_parsedFirstPart.emplace_back(std::move(result));
+    }
+    else
+    {
+        m_parsedFullName.emplace_back(std::move(result));
+    }
+}
+
+std::pair<LogLevel, bool> LogTagConfigParser::parseLogLevel(const std::string& s)
+{
+    const auto falseDontCare = std::make_pair(LOG_LEVEL_VERBOSE, false);
+    const size_t len = s.length();
+    if (len >= 1u)
+    {
+        std::pair<LogLevel, bool> result = falseDontCare;
+        // Fast classification based on first character.
+        // Tentative; need check remaining string.
+        char c = (char)std::toupper(s[0]);
+        switch (c)
+        {
+        case 'S':
+            result.first = LOG_LEVEL_SILENT;
+            result.second = (len == 1u || len == 6u);
+            break;
+        case 'F':
+            result.first = LOG_LEVEL_FATAL;
+            result.second = (len == 1u || len == 5u);
+            break;
+        case 'E':
+            result.first = LOG_LEVEL_ERROR;
+            result.second = (len == 1u || len == 5u);
+            break;
+        case 'W':
+            result.first = LOG_LEVEL_WARNING;
+            result.second = (len == 1u || len == 4u || len == 7u);
+            break;
+        case 'I':
+            result.first = LOG_LEVEL_INFO;
+            result.second = (len == 1u || len == 4u);
+            break;
+        case 'D':
+            result.first = LOG_LEVEL_DEBUG;
+            result.second = (len == 1u || len == 5u);
+            break;
+        case 'V':
+            result.first = LOG_LEVEL_VERBOSE;
+            result.second = (len == 1u || len == 7u);
+            break;
+        default:
+            break;
+        }
+        if (len == 1u)
+        {
+            return result;
+        }
+        if (!result.second)
+        {
+            return falseDontCare;
+        }
+        std::string upper = s;
+        std::transform(upper.begin(), upper.end(), upper.begin(),
+            [](char arg) -> char { return (char)toupper(arg); });
```

Comment:
The function 
```cpp
std::pair<LogLevel, bool> LogTagConfigParser::parseLogLevel(const std::string& s)
```
has been completely reworked to make it easier to understand and review for correctness. Even though the rewritten version may look like there are many calls to `toUpperCase`, these calls are mutually exclusive, so that any call to `parseLogLevel` will invoke `toUpperCase` at most once.